src/wp-includes/wp-db.php

```diff
@@ -1453,19 +1484,78 @@ public function prepare( $query, ...$args ) {
+		$query = preg_replace( "/%(?:%|$|(?!($allowed_format)?[sdfFi]))/", '%%\\1', $query ); // Escape any unescaped percents (i.e. anything unrecognised).
+
+		// Extract placeholders from the query.
+		$split_query = preg_split( "/(^|[^%]|(?:%%)+)(%(?:$allowed_format)?[sdfFi])/", $query, -1, PREG_SPLIT_DELIM_CAPTURE );
+
+		$split_query_count = count( $split_query );
+		$placeholder_count = ( ( $split_query_count - 1 ) / 3 ); // Split always returns with 1 value before the first placeholder (even with $query = "%s"), then 3 additional values per placeholder.
+
+		// If args were passed as an array (as in vsprintf), move them up.
+		$passed_as_array = ( isset( $args[0] ) && is_array( $args[0] ) && count( $args ) === 1 );
+		if ( $passed_as_array ) {
+			$args = $args[0];
+		}
+
+		$new_query       = '';
+		$key             = 2; // keys 0 and 1 in $split_query contain values before the first placeholder.
+		$arg_id          = 0;
+		$arg_identifiers = array();
+		$arg_strings     = array();
+		while ( $key < $split_query_count ) {
+
+			$placeholder = $split_query[ $key ];
+
+			$format = substr( $placeholder, 1, -1 );
+			$type   = substr( $placeholder, -1 );
 
-		$query = preg_replace( "/%(?:%|$|(?!($allowed_format)?[sdF]))/", '%%\\1', $query ); // Escape any unescaped percents.
+			if ( 'f' === $type ) { // Force floats to be locale-unaware.
+				$type        = 'F';
+				$placeholder = '%' . $format . $type;
+			}
 
-		// Count the number of valid placeholders in the query.
-		$placeholders = preg_match_all( "/(^|[^%]|(%%)+)%($allowed_format)?[sdF]/", $query, $matches );
+			if ( 'i' === $type ) {
+				$placeholder = '`%' . $format . 's`';
+				if ( strpos( $format, '$' ) !== false ) { // Using a simple strpos() due to previous checking (e.g. $allowed_format).
+					$arg_identifiers[] = intval( substr( $format, 1 ) );
+				} else {
+					$arg_identifiers[] = $arg_id;
+				}
+			} elseif ( 's' === $type ) {
+				if ( strpos( $format, '$' ) !== false ) {
+					$arg_strings[] = intval( substr( $format, 1 ) );
+				}
+				if ( true !== $this->unsafe_unquoted_parameters || '' === $format ) { // Unquoted strings for backwards compatibility (dangerous).
+					$placeholder = "'%" . $format . "s'";
+				}
+			} // elseif ( 'd' === $type || 'F' === $type ), nothing needs to be done (keep $placeholder)
+
+			$new_query .= $split_query[ $key - 2 ] . $split_query[ $key - 1 ] . $placeholder;
+				// Glue (-2), any prefix characters (-1), then the new $placeholder.
+
+			$key += 3;
+			$arg_id++;
+
+		}
+		$query = $new_query . $split_query[ $key - 2 ]; // Replace $query; and add remaining $query characters, or index 0 if there were no placeholders.
+
+		if ( count( array_intersect( $arg_identifiers, $arg_strings ) ) ) {
+			wp_load_translations_early();
+			_doing_it_wrong(
+				'wpdb::prepare',
+				__( 'You cannot use an argument for a string and an identifier.' ),
```

Comment:
Is this the error triggered by the `field --` unit test below? It's a pretty opaque error message.

Reply:
Yes, that error would be triggered by that test. I'm not too sure how to improve the text, but I've added the argument numbers to point developers in the right direction, and tweaked the wording a bit (further suggestions welcome).

Reply:
This was updated in commit [3b61b78](https://github.com/WordPress/wordpress-develop/pull/2192/commits/3b61b78466569b87c6cb6dc3dbe60e32c6f40cb5)